File: app/src/main/java/com/supercilex/robotscouter/ui/scout/ScoutPagerAdapter.java
package com.supercilex.robotscouter.ui.scout;

import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.text.TextUtils;
import android.view.View;
import android.widget.LinearLayout;

import com.google.firebase.crash.FirebaseCrash;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.RobotScouter;
import com.supercilex.robotscouter.data.util.ScoutUtils;
import com.supercilex.robotscouter.data.util.TeamHelper;

import java.util.ArrayList;
import java.util.List;

import static com.supercilex.robotscouter.util.ConnectivityUtilsKt.isOffline;
import static com.supercilex.robotscouter.util.ConstantsKt.FIREBASE_NAME;
import static com.supercilex.robotscouter.util.ConstantsKt.FIREBASE_SCOUTS;
import static com.supercilex.robotscouter.util.ConstantsKt.SINGLE_ITEM;

public class ScoutPagerAdapter extends FragmentStatePagerAdapter
        implements ValueEventListener, TabLayout.OnTabSelectedListener, View.OnLongClickListener {
    private final ValueEventListener mTabNameListener = new ValueEventListener() {
        @Override
        public void onDataChange(DataSnapshot snapshot) {
            int tabIndex = mKeys.indexOf(snapshot.getRef().getParent().getKey());
            TabLayout.Tab tab = mTabLayout.getTabAt(tabIndex);
            tab.setText(snapshot.getValue() == null ?
                                getPageTitle(tabIndex) : snapshot.getValue(String.class));
            View tabView = ((LinearLayout) mTabLayout.getChildAt(0)).getChildAt(tabIndex);
            tabView.setOnLongClickListener(ScoutPagerAdapter.this);
            tabView.setId(tabIndex);
        }

        @Override
        public void onCancelled(DatabaseError error) {
            ScoutPagerAdapter.this.onCancelled(error);
        }
    };

    private final List<String> mKeys = new ArrayList<>();

    private final Fragment mFragment;
    private final AppBarViewHolderBase mAppBarViewHolder;
    private final TabLayout mTabLayout;
    private final TeamHelper mTeamHelper;
    private final Query mQuery;

    private String mCurrentScoutKey;

    public ScoutPagerAdapter(Fragment fragment,
                             AppBarViewHolderBase appBarViewHolder,
                             TabLayout tabLayout,
                             TeamHelper helper,
                             String currentScoutKey) {
        super(fragment.getChildFragmentManager());
        mFragment = fragment;
        mAppBarViewHolder = appBarViewHolder;
        mTabLayout = tabLayout;
        mTeamHelper = helper;
        mCurrentScoutKey = currentScoutKey;
        mQuery = ScoutUtils.getIndicesRef(helper.getTeam().getKey());
        mQuery.addValueEventListener(this);
    }

    @Override
    public Fragment getItem(int position) {
        return ScoutFragment.newInstance(mKeys.get(position));
    }

    @Override
    public int getCount() {
        return mKeys.size();
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return "Scout " + (getCount() - position);
    }

    public String getCurrentScoutKey() {
        return mCurrentScoutKey;
    }

    public void setCurrentScoutKey(String currentScoutKey) {
        mCurrentScoutKey = currentScoutKey;
    }

    public void onScoutDeleted() {
        int index = mKeys.indexOf(mCurrentScoutKey);
        String newKey = null;
        if (mKeys.size() > SINGLE_ITEM) {
            newKey = mKeys.size() - 1 > index ? mKeys.get(index + 1) : mKeys.get(index - 1);
        }
        ScoutUtils.delete(mQuery.getRef().getKey(), mCurrentScoutKey);
        mCurrentScoutKey = newKey;
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        mCurrentScoutKey = mKeys.get(tab.getPosition());
    }

    @Override
    public void onDataChange(DataSnapshot snapshot) {
        removeNameListeners();
        boolean hadScouts = !mKeys.isEmpty();
        mKeys.clear();
        for (DataSnapshot scoutIndex : snapshot.getChildren()) {
            String key = scoutIndex.getKey();
            mKeys.add(0, key);
            getTabNameRef(key).addValueEventListener(mTabNameListener);
        }
        if (hadScouts
                && mKeys.isEmpty()
                && !isOffline(mFragment.getContext())
                && mFragment.isResumed()) {
            ShouldDeleteTeamDialog.Companion.show(mFragment.getChildFragmentManager(), mTeamHelper);
        }
        mFragment.getView().findViewById(R.id.no_content_hint)
                .setVisibility(mKeys.isEmpty() ? View.VISIBLE : View.GONE);
        mAppBarViewHolder.setDeleteScoutMenuItemVisible(!mKeys.isEmpty());

        mTabLayout.removeOnTabSelectedListener(this);
        notifyDataSetChanged();
        mTabLayout.addOnTabSelectedListener(this);

        if (!mKeys.isEmpty()) {
            if (TextUtils.isEmpty(mCurrentScoutKey)) {
                selectTab(0);
                mCurrentScoutKey = mKeys.get(0);
            } else {
                selectTab(mKeys.indexOf(mCurrentScoutKey));
            }
        }
    }

    private void selectTab(int index) {
        TabLayout.Tab tab = mTabLayout.getTabAt(index);
        if (tab != null) tab.select();
    }

    private DatabaseReference getTabNameRef(String key) {
        return FIREBASE_SCOUTS.child(key).child(FIREBASE_NAME);
    }

    public void cleanup() {
        mQuery.removeEventListener(this);
        removeNameListeners();
        RobotScouter.Companion.getRefWatcher(mFragment.getActivity()).watch(this);
    }

    private void removeNameListeners() {
        for (String key : mKeys) getTabNameRef(key).removeEventListener(mTabNameListener);
    }

    @Override
    public boolean onLongClick(View v) {
        ScoutNameDialog.Companion.show(
                mFragment.getChildFragmentManager(),
                FIREBASE_SCOUTS.child(mKeys.get(v.getId())).child(FIREBASE_NAME),
                mTabLayout.getTabAt(v.getId()).getText().toString());
        return true;
    }

    @Override
    public void onCancelled(DatabaseError error) {
        FirebaseCrash.report(error.toException());
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {
        // Not interesting
    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {
        // Not interesting
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scout/ScoutPagerAdapter.kt
package com.supercilex.robotscouter.ui.scout

import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleFragment
import android.arch.lifecycle.LifecycleObserver
import android.arch.lifecycle.Observer
import android.arch.lifecycle.OnLifecycleEvent
import android.arch.lifecycle.ViewModelProviders
import android.support.design.widget.TabLayout
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.PagerAdapter
import android.text.TextUtils
import android.view.View
import android.widget.LinearLayout
import com.google.firebase.crash.FirebaseCrash
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.RobotScouter
import com.supercilex.robotscouter.data.model.Team
import com.supercilex.robotscouter.data.util.deleteScout
import com.supercilex.robotscouter.util.FIREBASE_NAME
import com.supercilex.robotscouter.util.FIREBASE_SCOUTS
import com.supercilex.robotscouter.util.SINGLE_ITEM
import com.supercilex.robotscouter.util.isOffline
import java.util.ArrayList

class ScoutPagerAdapter(private val fragment: LifecycleFragment,
                        private val appBarViewHolder: AppBarViewHolderBase,
                        private val tabLayout: TabLayout,
                        private val team: Team,
                        var currentScoutKey: String?) :
        FragmentStatePagerAdapter(fragment.childFragmentManager),
        Observer<List<String>>, TabLayout.OnTabSelectedListener, View.OnLongClickListener, LifecycleObserver {
    private val tabNameListener = object : ValueEventListener {
        override fun onDataChange(snapshot: DataSnapshot) {
            val tabIndex = keys.indexOf(snapshot.ref.parent.key)
            val tab = tabLayout.getTabAt(tabIndex)

            tab!!.text = if (snapshot.value == null)
                getPageTitle(tabIndex)
            else
                snapshot.getValue(String::class.java)

            val tabView = (tabLayout.getChildAt(0) as LinearLayout).getChildAt(tabIndex)
            tabView.setOnLongClickListener(this@ScoutPagerAdapter)
            tabView.id = tabIndex
        }

        override fun onCancelled(error: DatabaseError) = FirebaseCrash.report(error.toException())
    }

    private val holder: ScoutKeysHolder = ViewModelProviders.of(fragment).get(ScoutKeysHolder::class.java)
    private var keys: List<String> = ArrayList()

    init {
        holder.init(team.key)
        holder.keysListener.observe(fragment, this)

        fragment.lifecycle.addObserver(this)
    }

    override fun getItem(position: Int) = ScoutFragment.newInstance(keys[position])

    override fun getCount() = keys.size

    override fun getItemPosition(any: Any?) = PagerAdapter.POSITION_NONE

    override fun getPageTitle(position: Int): String =
            fragment.getString(R.string.title_scout_tab, count - position)

    fun onScoutDeleted() {
        val index = keys.indexOf(currentScoutKey)
        var newKey: String? = null
        if (keys.size > SINGLE_ITEM) {
            newKey = if (keys.size - 1 > index) keys[index + 1] else keys[index - 1]
        }
        deleteScout(holder.ref.key, currentScoutKey!!)
        currentScoutKey = newKey
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        currentScoutKey = keys[tab.position]
    }

    override fun onChanged(newKeys: List<String>?) {
        removeNameListeners()
        val hadScouts = !keys.isEmpty()

        keys = newKeys!!

        for (key in keys) getTabNameRef(key).addValueEventListener(tabNameListener)
        if (hadScouts && keys.isEmpty() && !isOffline() && fragment.isResumed) {
            ShouldDeleteTeamDialog.show(fragment.childFragmentManager, team)
        }
        fragment.view!!.findViewById<View>(R.id.no_content_hint).visibility =
                if (keys.isEmpty()) View.VISIBLE else View.GONE
        appBarViewHolder.setDeleteScoutMenuItemVisible(!keys.isEmpty())

        tabLayout.removeOnTabSelectedListener(this)
        notifyDataSetChanged()
        tabLayout.addOnTabSelectedListener(this)

        if (!keys.isEmpty()) {
            if (TextUtils.isEmpty(currentScoutKey)) {
                selectTab(0)
                currentScoutKey = keys[0]
            } else {
                selectTab(keys.indexOf(currentScoutKey))
            }
        }
    }

    private fun selectTab(index: Int) = tabLayout.getTabAt(index)?.select()

    private fun getTabNameRef(key: String) = FIREBASE_SCOUTS.child(key).child(FIREBASE_NAME)

    @Suppress("unused")
    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    private fun cleanup() {
        fragment.lifecycle.removeObserver(this)
        removeNameListeners()
        RobotScouter.getRefWatcher(fragment.activity).watch(this)
    }

    private fun removeNameListeners() {
        for (key in keys) getTabNameRef(key).removeEventListener(tabNameListener)
    }

    override fun onLongClick(v: View): Boolean {
        ScoutNameDialog.show(
                fragment.childFragmentManager,
                FIREBASE_SCOUTS.child(keys[v.id]).child(FIREBASE_NAME),
                tabLayout.getTabAt(v.id)!!.text!!.toString())
        return true
    }

    override fun onTabUnselected(tab: TabLayout.Tab) = Unit

    override fun onTabReselected(tab: TabLayout.Tab) = Unit
}
